### PR TITLE
fix checkpointing of non bonded interactions

### PR DESCRIPTION
**Pull request by KaiSzuttor**
… having a system instance.

### CODE REVIEW

**Comment by KaiSzuttor** [comment]
~stil~l non_bonded_inter can~not~ be recovered correctly:

``` python
from __future__ import print_function                                                                       
import numpy as np
import espressomd
import espressomd.checkpointing

checkpoint = espressomd.checkpointing.Checkpointing(checkpoint_id="checkpoint")


if not checkpoint.has_checkpoints():
    print("Starting from scratch")
    system = espressomd.System(box_l = [10,10,10])
    checkpoint.register("system")
    system.cell_system.skin = 0.1
    system.time_step = 0.01
    for i in range(10):
        system.part.add(pos=system.box_l * np.random.random(3))
    system.non_bonded_inter[0, 1].lennard_jones.set_params(sigma=1.0, epsilon=1.0, shift='auto', cutoff=2.3)
    print(system.non_bonded_inter[0, 1].lennard_jones)

restarted = False
if checkpoint.has_checkpoints():
    print("Restoring Equilibration Checkpoint")
    checkpoint.load()
    print(system.non_bonded_inter[0, 1].lennard_jones)
    print("Checkpoint loaded")
    restarted = True
```

File: src/core/interaction_data.cpp [hunk 1]
```diff
@@ -162,6 +163,7 @@ void ia_params_set_state(std::string const &state) {
   boost::archive::binary_iarchive ia(ss);
   ia_params.clear();
   ia >> ia_params;
+  ia >> max_seen_particle_type;
```
Comment (fweik) [inline src/core/interaction_data.cpp:166]: I think you have to set max_seen_particle_type to the maximum of its current value and the value from the archive, as it's current value could be higher than the stored one.

**Review by fweik — CHANGES_REQUESTED** [review]

File: src/core/interaction_data.cpp (hunk 1 above)
Comment (KaiSzuttor) [reply src/core/interaction_data.cpp:166]: but then I'm also not allowed to clear ```ia_params```, somebody could have created interactions before loading the checkpoint

Comment (KaiSzuttor) [reply src/core/interaction_data.cpp:166]: in my opinion the user cannot expect that espresso is able to merge current state with saved state

File: src/core/interaction_data.cpp [hunk 2]
```diff
@@ -162,6 +163,9 @@ void ia_params_set_state(std::string const &state) {
...
+  mpi_bcast_max_seen_particle_type(ia_params.size());
```
Comment (fweik) [inline src/core/interaction_data.cpp]: I don't think this is correct, ia_params contains all type pairs.

**Review by fweik — CHANGES_REQUESTED** [review]

**Comment by KaiSzuttor** [comment]
I just noticed that it's still not possible to correctly load a system instance without having created a system instance before. The checkpoint test now fails for the particle position. I don't know why yet.

**Comment by KaiSzuttor** [comment]
The following problem arose during this PR:
We have a circular dependency that cannot be overcome without changing the core:
 - CellSystem instance is constructed in Systems constructor
 - CellSystem inherently depends on the box geometry
 - The box geometry is set by the System constructor
 - System instance depends on CellSystem instance

**Comment by KaiSzuttor** [comment]
Similar problem arises for the dependency of ParticleList on the global variable ```time_step```. See failing testcase.

**Comment by fweik** [comment]
So there is no solution for this on the python level?

**Comment by KaiSzuttor** [comment]
if we introduce an object instance that sets all globals and is initialized by pickle first, I think that should work

**Comment by hmenke** [comment]
I haven't been following the changes in the repo too closely in the recent past.  What has changed that makes checkpointing such a big issue suddenly?  I remember that it used to work just fine.

**Comment by KaiSzuttor** [comment]
its not a big issue, I'm trying to make it more comfortable by making it possible to checkpoint "system" instead of single components.

**Comment by hmenke** [comment]
I admit, my comment is not well-placed in this thread.  Actually I was more referring to #1938, #1960, and #1969.

**Comment by KaiSzuttor** [comment]
@fweik introducing an object that sets global variables would lead to an interface change, e.g.
``` python
system.globals.box_l = [10.0, 10.0, 10.0]
```
do we want that just because of checkpointing issues?

**Comment by RudolfWeeber** [comment]
On Fri, May 25, 2018 at 03:00:52AM -0700, Kai Szuttor wrote:
> @fweik introducing an object that sets global variables would lead to an interface change, e.g.
> ``` python
> system.globals.box_l = [10.0, 10.0, 10.0]
> ```
> do we want that just because of checkpointing issues?
I'd say no. It's cumbersome and what is or isn't a 'global' in the core shouldn't be of concern to the user.

**Comment by fweik** [comment]
Can't you get that with a property without an interface change?

**Review by fweik — APPROVED** [review]

**Pull request merged**